play-json/shared/src/main/scala/play/api/libs/json/JsPath.scala

```diff
@@ -63,24 +63,17 @@ case class KeyPathNode(key: String) extends PathNode {
 
   def set(json: JsValue, transform: JsValue => JsValue): JsValue = json match {
     case obj: JsObject =>
-      var found = false
-      val o = JsObject(obj.fields.map { case (k, v) =>
-        if (k == this.key) {
-          found = true
-          k -> transform(v)
-        } else k -> v
-      })
-      if (!found) o ++ JsObject(Seq(this.key -> transform(JsObject.empty)))
-      else o
+      val transformed = transform(obj.underlying.getOrElse(key, JsObject.empty))
+      obj + (key -> transformed)
```

Comment:
Note that our underlying `LinkedHashMap` preserves order on `+` for existing keys.
```scala
scala> lhm
val res3: java.util.LinkedHashMap[String, Int] = {1=1, 2=2, 3=3}

scala> lhm.put("2", 2222222)
val res4: Int = 2

scala> lhm
val res5: java.util.LinkedHashMap[String, Int] = {1=1, 2=2222222, 3=3}
```